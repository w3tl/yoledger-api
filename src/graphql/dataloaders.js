import DataLoader from 'dataloader';
import { Account } from '../model';

const getAccountByName = connection => ids => new Promise((resolve, reject) => {
  const promises = ids.map(
    ({ name, type, userId }) => {
      const account = new Account(connection.db(), userId);
      return account.findOne({ name, type });
    },
  );
  Promise.all(promises)
    .then(docs => resolve(docs))
    .catch(err => reject(err));
});

export default connection => ({
  accountByName: new DataLoader(getAccountByName(connection)),
});
