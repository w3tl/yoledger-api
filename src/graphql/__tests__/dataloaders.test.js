jest.mock('mongodb');
import dataloaders from '../dataloaders';

describe('dataloaders', () => {
  const { connection } = require('mongodb');
  const dl = dataloaders(connection);

  test('accountByName', async () => {
    const load = await dl.accountByName.load({ name: 'Food', type: 'EXPENSE', userId: 'admin' });
    expect(load).toMatchSnapshot('Load Food for admin');

    const loadMany = await dl.accountByName.loadMany([
      { name: 'Food', type: 'EXPENSE', userId: 'admin' },
      { name: 'Train', type: 'EXPENSE', userId: 'admin' },
    ]);
    expect(loadMany).toMatchSnapshot('Load Food and Train for admin');
  });
});
