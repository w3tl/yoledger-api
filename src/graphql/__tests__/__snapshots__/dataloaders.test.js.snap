// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataloaders accountByName: Load Food and Train for admin 1`] = `
Array [
  Object {
    "_id": "3",
    "name": "Food",
    "type": "EXPENSE",
  },
  Object {
    "_id": "4",
    "name": "Train",
    "type": "EXPENSE",
  },
]
`;

exports[`dataloaders accountByName: Load Food for admin 1`] = `
Object {
  "_id": "3",
  "name": "Food",
  "type": "EXPENSE",
}
`;
